src/mscorlib/src/System/IO/UnmanagedMemoryStream.cs

```diff
@@ -404,27 +405,34 @@ protected override void Dispose(bool disposing)
 
             int nInt = (int) n; // Safe because n <= count, which is an Int32
             if (nInt < 0)
-                nInt = 0;  // _position could be beyond EOF
+                return 0;  // _position could be beyond EOF
             Contract.Assert(pos + nInt >= 0, "_position + n >= 0");  // len is less than 2^63 -1.
 
-            if (_buffer != null) {
-                unsafe {
-                    byte* pointer = null;
-                    RuntimeHelpers.PrepareConstrainedRegions();
-                    try {
-                        _buffer.AcquirePointer(ref pointer);
-                        Buffer.Memcpy(buffer, offset, pointer + pos + _offset, 0, nInt);
-                    }
-                    finally {
-                        if (pointer != null) {
-                            _buffer.ReleasePointer();
+            unsafe
+            {
+                fixed (byte* pBuffer = buffer)
+                {
+                    if (_buffer != null)
+                    {
+                        byte* pointer = null;
+
+                        try
+                        {
+                            _buffer.AcquirePointer(ref pointer);
+                            Buffer.MemoryCopy(source: pointer + pos + _offset, destination: pBuffer + offset, destinationSizeInBytes: nInt, sourceBytesToCopy: nInt);
```

Comment:
The original code that called `Buffer.Memcpy` is better. `Buffer.MemoryCopy` is wrapper around MemCpy with unnecessary overhead.